[22.05] Fix bulk operations visibility

davelopez opened the pull request:
Fixes #14107

- The `Permanently delete` option now appears always.
- The `Undelete` option now is hidden when using manual selection (in query selection we cannot know if there would be non-purged datasets selected) and all items are purged.

**Note**: Trying to purge or undelete a purged dataset will now raise an error in the API. This will probably trigger more error dialogs in the client due to *partial success* but it will make it easier to detect when the history will actually change after an operation so we can stop the `Processing...` message correctly.

![fix_bulk_ops_visibility](https://user-images.githubusercontent.com/46503462/174632456-6f3e782e-8bcf-4239-83d3-1ee252c4c8c7.gif)


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Follow instructions on #14107

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Always display `purge` option in bulk

Regardless the datasets are deleted or not.

davelopez committed:
Raise error when bulk undeleting

Instead of doing nothing and treating the item as successful, we now raise a specific error.
This helps to determine if the bulk operation will trigger a history update when an item succeeds.
Otherwise, the UI will keep waiting for a history update that may never occur.
For the same reason, when purging an already purged item, we will touch the `update_time` but do nothing, so we notify the history to expect a *change*.

davelopez committed:
Reduce error dialog dramatism when partial success

Now the title "Something went wrong..." will be visible only when all items fail.

davelopez committed:
Hide bulk undelete when all selected items purged

When we are in query selection mode we need to enable it anyway since we don't know if some items may not be purged in the query.

nsoranzo commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1602,11 +1602,13 @@ def _delete(self, item: HistoryItemModel):
 
     def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        if getattr(item, "purged", False):
+            raise exceptions.ItemDeletionException("This item was already permanently deleted.")
```
Maybe this can just be a no-op?
```suggestion
            return
```

davelopez replied:
This is intentional, although it might be a better solution, see the *note* in the PR description. Your suggestion was the initial approach, but there is a corner case...
If we select only purged datasets and then try to purge them in bulk (for whatever reason) if we no-op, the API call will return success for all of them in the response, and then the UI will expect a history update with the updated items... which will never occur and we get the `Processing operation...` display forever or until page refresh...

So in summary, if we no-op it's hard to know if we should wait for history changes or not, this way since we know there will be no change, we can drop the processing message at the cost of displaying an error which is kind of true :sweat_smile:

mvdbeek replied:
Sounds like we need to track the progress of a batch operation, it seems unfortunate that we have to raise an exception here to work around not having a way to do this currently. Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

davelopez replied:
> Sounds like we need to track the progress of a batch operation

Yes, that would be the ideal solution.

> Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Sure, that sounds more reasonable, thanks for the suggestion, will do!

davelopez replied:
By the way, is it ok to raise the error when trying to `undelete` a `purged` dataset or should we do the same, no-op, and just touch the `update_time`?

mvdbeek replied:
No, I'd raise an exception there.

davelopez replied:
Done! Thanks for the reviews!

davelopez committed:
Fix unit test

Fixes:
[vue-test-utils]: wrapper.setProps() called with the same object of the existing contentSelection property. You must call wrapper.setProps() with a new object to trigger reactivity
